platform/configurator/configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ManagedServiceOperation.java

```diff
@@ -0,0 +1,197 @@
+import org.codice.ddf.admin.configurator.Operation;
+import org.codice.ddf.ui.admin.api.ConfigurationAdmin;
+import org.codice.ddf.ui.admin.api.ConfigurationAdminMBean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Transactional handler factory for creating and deleting managed services.
+ * <p>
+ * <b> This code is experimental. While this class is functional and tested, it may change or be
+ * removed in a future version of the library. </b>
+ */
+public abstract class ManagedServiceOperation
+        implements OperationBase, Operation<String, Map<String, Map<String, Object>>> {
+    /**
+     * Transactional handler for deleting managed services.
+     */
+    private static class DeleteHandler extends ManagedServiceOperation {
+        private final String configPid;
+
+        private Map<String, Object> currentProperties;
+
+        private DeleteHandler(String configPid, ConfigurationAdmin configAdmin,
+                ConfigurationAdminMBean cfgAdmMbean) {
+            super(configAdmin, cfgAdmMbean);
+
+            this.configPid = configPid;
+
+            try {
+                factoryPid = cfgAdmMbean.getFactoryPid(configPid);
+                currentProperties = cfgAdmMbean.getProperties(configPid);
+            } catch (IOException e) {
+                ManagedServiceOperation.LOGGER.debug(
+                        "Error getting current configuration for pid {}",
+                        configPid,
+                        e);
+                throw new ConfiguratorException("Internal error");
+            }
+        }
+
+        @Override
+        public String commit() throws ConfiguratorException {
+            deleteByPid(configPid);
+            return null;
+        }
+
+        @Override
+        public String rollback() throws ConfiguratorException {
+            return createManagedService(currentProperties);
+        }
+    }
+
+    /**
+     * Transactional handler for creating managed services.
+     */
+    private static class CreateHandler extends ManagedServiceOperation {
+        private final Map<String, Object> configs;
+
+        private String newConfigPid;
+
+        private CreateHandler(String factoryPid, Map<String, Object> configs,
+                ConfigurationAdmin configAdmin, ConfigurationAdminMBean cfgAdmMbean) {
+            super(configAdmin, cfgAdmMbean);
+
+            this.factoryPid = factoryPid;
+            this.configs = new HashMap<>(configs);
+        }
+
+        @Override
+        public String commit() throws ConfiguratorException {
+            newConfigPid = createManagedService(configs);
+            return newConfigPid;
+        }
+
+        @Override
+        public String rollback() throws ConfiguratorException {
+            deleteByPid(newConfigPid);
+            return null;
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ManagedServiceOperation.class);
+
+    protected String factoryPid;
```

Comment:
The problem with moving the `factoryPid` initialization to the base class is that in the case of the `DeleteHandler` implementor I have to query to find the `factoryPid` and the query can throw a checked exception. I could get around that but it would be a lot of extra code that I’m not sure I see the value of in an internal implementation.

I will play around with it a bit more to see if I can come up with something that's not overly complex, but right now my feeling is that the exposure is low.